File: backend/app/services/insights_service.py
from typing import List, Dict, Any
from collections import Counter
import re
from app.models.schemas import AIInsights

REQUEST_PATTERNS = re.compile(
    r"\b(please|can you|could you|more|next|part\s*\d+|tutorial|"
    r"explain|show|make|need|want|wish|hope)\b",
    re.IGNORECASE,
)


class InsightsService:
    def _top_keywords_for_sentiment(
        self,
        comments: List[Dict[str, Any]],
        sentiment: str,
        n: int = 10,
    ) -> List[str]:
        stop = {"the", "a", "an", "is", "it", "in", "on", "to", "for", "and", "of",
                "this", "that", "was", "are", "you", "i", "my", "your", "so", "just",
                "very", "too", "its", "like", "im", "but", "not"}
        words: List[str] = []
        for c in comments:
            if c.get("sentiment") == sentiment:
                words.extend(
                    w for w in re.findall(r"\b[a-z]{3,}\b", c["text"].lower())
                    if w not in stop
                )
        counter = Counter(words)
        return [w for w, _ in counter.most_common(n)]

    def _extract_requests(self, comments: List[Dict[str, Any]]) -> List[str]:
        requests = []
        for c in comments:
            if REQUEST_PATTERNS.search(c["text"]):
                requests.append(c["text"][:120].strip())
        return requests[:5]

    def generate_insights(
        self,
        comments: List[Dict[str, Any]],
        sentiment_dist: Dict[str, Any],
    ) -> AIInsights:
        pos_keywords = self._top_keywords_for_sentiment(comments, "positive")
        neg_keywords = self._top_keywords_for_sentiment(comments, "negative")
        requests = self._extract_requests(comments)

        total = len(comments)
        pos_pct = sentiment_dist.get("positive_pct", 0)
        neg_pct = sentiment_dist.get("negative_pct", 0)

        liked_most = (
            f"Viewers reacted positively to: {', '.join(pos_keywords[:6])}. "
            f"{pos_pct}% of comments expressed positive sentiment."
            if pos_keywords
            else "No strong positive themes detected."
        )

        disliked_most = (
            f"Common negative themes: {', '.join(neg_keywords[:6])}. "
            f"{neg_pct}% of comments were negative."
            if neg_keywords
            else "Very few negative comments detected."
        )

        common_requests = (
            "Top viewer requests: " + " | ".join(f'"{r}"' for r in requests[:3])
            if requests
            else "No clear viewer requests detected in comments."
        )

        if pos_pct >= 60:
            tone = "overwhelmingly positive"
            suggestion = "Continue creating similar content. Consider a follow-up or series."
        elif pos_pct >= 40:
            tone = "generally positive with room for improvement"
            suggestion = "Address the negative feedback themes and engage with viewer questions."
        else:
            tone = "mixed to negative"
            suggestion = "Review content quality, pacing, or topic alignment with your audience."

        improvement_suggestions = (
            f"Audience sentiment is {tone}. {suggestion} "
            f"{'Top requests: ' + ', '.join(requests[:2]) if requests else ''}"
        ).strip()

        overall_summary = (
            f"Analyzed {total} comments. {pos_pct}% positive, "
            f"{neg_pct}% negative. Audience tone is {tone}."
        )

        return AIInsights(
            liked_most=liked_most,
            disliked_most=disliked_most,
            common_requests=common_requests,
            improvement_suggestions=improvement_suggestions,
            overall_sentiment_summary=overall_summary,
        )


insights_service = InsightsService()

File: backend/src/services/insightsService.js
const REQUEST_PATTERN = /\b(please|can you|could you|more|next|part\s*\d+|tutorial|explain|show|make|need|want|wish|hope)\b/i;

const INSIGHT_STOP = new Set([
  'the', 'a', 'an', 'is', 'it', 'in', 'on', 'to', 'for', 'and', 'of',
  'this', 'that', 'was', 'are', 'you', 'i', 'my', 'your', 'so', 'just',
  'very', 'too', 'its', 'like', 'im', 'but', 'not',
]);

class InsightsService {
  _topKeywordsForSentiment(comments, sentiment, n = 10) {
    const wordCounts = {};
    for (const c of comments) {
      if (c.sentiment !== sentiment) continue;
      const words = c.text.toLowerCase().match(/\b[a-z]{3,}\b/g) || [];
      for (const w of words) {
        if (!INSIGHT_STOP.has(w)) {
          wordCounts[w] = (wordCounts[w] || 0) + 1;
        }
      }
    }
    return Object.entries(wordCounts)
      .sort((a, b) => b[1] - a[1])
      .slice(0, n)
      .map(([w]) => w);
  }

  _extractRequests(comments) {
    return comments
      .filter(c => REQUEST_PATTERN.test(c.text))
      .slice(0, 5)
      .map(c => c.text.slice(0, 120).trim());
  }

  generateInsights(comments, sentimentDist) {
    const posKeywords = this._topKeywordsForSentiment(comments, 'positive');
    const negKeywords = this._topKeywordsForSentiment(comments, 'negative');
    const requests = this._extractRequests(comments);

    const total = comments.length;
    const pos_pct = sentimentDist.positive_pct || 0;
    const neg_pct = sentimentDist.negative_pct || 0;

    const liked_most = posKeywords.length
      ? `Viewers reacted positively to: ${posKeywords.slice(0, 6).join(', ')}. ${pos_pct}% of comments expressed positive sentiment.`
      : 'No strong positive themes detected.';

    const disliked_most = negKeywords.length
      ? `Common negative themes: ${negKeywords.slice(0, 6).join(', ')}. ${neg_pct}% of comments were negative.`
      : 'Very few negative comments detected.';

    const common_requests = requests.length
      ? 'Top viewer requests: ' + requests.slice(0, 3).map(r => `"${r}"`).join(' | ')
      : 'No clear viewer requests detected in comments.';

    let tone, suggestion;
    if (pos_pct >= 60) {
      tone = 'overwhelmingly positive';
      suggestion = 'Continue creating similar content. Consider a follow-up or series.';
    } else if (pos_pct >= 40) {
      tone = 'generally positive with room for improvement';
      suggestion = 'Address the negative feedback themes and engage with viewer questions.';
    } else {
      tone = 'mixed to negative';
      suggestion = 'Review content quality, pacing, or topic alignment with your audience.';
    }

    const topRequests = requests.slice(0, 2);
    const improvement_suggestions = (
      `Audience sentiment is ${tone}. ${suggestion}` +
      (topRequests.length ? ` Top requests: ${topRequests.join(', ')}` : '')
    ).trim();

    const overall_sentiment_summary =
      `Analyzed ${total} comments. ${pos_pct}% positive, ${neg_pct}% negative. Audience tone is ${tone}.`;

    return {
      liked_most,
      disliked_most,
      common_requests,
      improvement_suggestions,
      overall_sentiment_summary,
    };
  }
}

module.exports = new InsightsService();
